src/noit_metric_tag_search.c

```diff
@@ -1332,34 +1332,15 @@ noit_metric_tag_search_evaluate_against_metric_id(const noit_metric_tag_search_a
 
   // setup check tags
   noit_metric_tagset_t tagset_check = id->check;
-  // Add in extra tags: __uuid
-  if (tagset_check.tag_count > MAX_TAGS - 1) {
-    mtev_memory_end();
-    return 0;
-  }
   MKTAGSETCOPY(tagset_check);
-  char uuid_str[13 + UUID_STR_LEN + 1];
-  strcpy(uuid_str, "__check_uuid:");
-  mtev_uuid_unparse_lower(id->id, uuid_str + 13);
-  noit_metric_tag_t uuid_tag = { .tag = uuid_str, .total_size = strlen(uuid_str), .category_size = 13 };
-  tagset_check.tags[tagset_check.tag_count++] = uuid_tag;
   if(noit_metric_tagset_fixup_hook_invoke(NOIT_METRIC_TAGSET_CHECK, &tagset_check) == MTEV_HOOK_ABORT) {
```

Comment:
This invokes snowth's `check_tag_tagset_completer`, which will always return `MTEV_HOOK_CONTINUE`.
When `NOIT_METRIC_TAGSET_CHECK` is used, the function will check for the `__check_uuid` tag and log to `use_debug` if it's not found (all other options will just return immediately). 
In this case, it will never find the `__check_uuid` because it has been moved to `tagset_implicit` to allow `tagset_check` to utilize its max potential of `MAX_TAGS` tags instead of `MAX_TAGS -1` tags.
We use `NOIT_METRIC_TAGSET_CHECK` to correctly check for the `__check_uuid` tag when running this against the `tagset_implicit` tagset, but I am not sure what to do for this call for the `tagset_check` tagset.

The other options for `noit_metric_tagset_class_t` are `NOIT_METRIC_TAGSET_STREAM` and `NOIT_METRIC_TAGSET_MEASUREMENT`, which would be inappropriate to use here, even though they don't perform any logic in the function registered in snowth.

On the other hand, leaving this to use `NOIT_METRIC_TAGSET_CHECK` doesn't feel right either, because we are unnecessarily checking for the existence of the `__check_uuid` tag that we know will never be in the tagset (and logging the "missing" tag to `use_debug' every time).

We could maybe just not invoke this function for `tagset_check`, but I don't know how other projects may be registering this, so I don't want to assume other implementations work the same as snowth.


The cleanest solution is to create a new enum called `NOIT_METRIC_TAGSET_IMPLICIT` and change snowth's `check_tag_tagset_completer` to only perform the `__check_uuid` logic for the new enum (instead of the CHECK enum that it currently performs it for), but introducing a new enum to reconnoiter (and changing the existing logic surrounding them) may break other implementations of `noit_metric_tagset_fixup_hook_invoke`.
I have yet to check if we have any other implementation of this, but the project is open-source, so I'd like to discuss any of these changes before I make them on my own, even if snowth is circonus's only implementation of this function.

There's also the timing issue of deploying these changes to both reconnoiter and snowth in a way that does not break the functionality (maybe perform the `__check_uuid` logic on both CHECK and IMPLICIT enums (and sacrifice the extra processing and logs) until we are sure all libnoits have been updated and are comfortable updating snowth to only perform the logic on the IMPLICIT enum.

Reply:
> This invokes snowth's check_tag_tagset_completer, which will always return MTEV_HOOK_CONTINUE.
When NOIT_METRIC_TAGSET_CHECK is used, the function will check for the __check_uuid tag and log to use_debug if it's not found (all other options will just return immediately).
In this case, it will never find the __check_uuid because it has been moved to tagset_implicit to allow tagset_check to utilize its max potential of MAX_TAGS tags instead of MAX_TAGS -1 tags.
We use NOIT_METRIC_TAGSET_CHECK to correctly check for the __check_uuid tag when running this against the tagset_implicit tagset, but I am not sure what to do for this call for the tagset_check tagset.

It seems to be also null terminating all the uuid strings.  This behavior should be retained and verified

> The cleanest solution is to create a new enum called NOIT_METRIC_TAGSET_IMPLICIT and change snowth's check_tag_tagset_completer to only perform the __check_uuid logic for the new enum (instead of the CHECK enum that it currently performs it for), but introducing a new enum to reconnoiter (and changing the existing logic surrounding them) may break other implementations of noit_metric_tagset_fixup_hook_invoke.
I have yet to check if we have any other implementation of this, but the project is open-source, so I'd like to discuss any of these changes before I make them on my own, even if snowth is circonus's only implementation of this function.

don't understand what that means

Reply:
There was no need to change anything outside of this function. I changed this back to adding the implicit tags directly to the tagsets already being created in this scope, but made changes to allow this to work even if the MAX_TAGS are already being used. Since this is copying the tags to a local tagset, we can just make that local tagset size MAX_TAGS+1 (there are no restrictions within the eval function called on these tagsets that would actually restrict the tag count to MAX_TAGS).